edb/schema/constraints.py

```diff
@@ -176,18 +176,27 @@ def get_name_impacting_ancestors(
             return [self.get_nearest_generic_parent(schema)]
 
     def get_constraint_origins(
-            self, schema: s_schema.Schema) -> List[Constraint]:
-        origins: List[Constraint] = []
+        self, schema: s_schema.Schema
+    ) -> List[Constraint]:
+        """
+        Find the transitive closure of all constraints that apply.
+        Return parents of these constraints.
+        """
+        # I'm not sure that this explanation is clear (or even entirely correct)
+
+        origins: Set[Constraint] = set()
         for base in self.get_bases(schema).objects(schema):
-            if not base.is_non_concrete(schema) and not base.get_delegated(
-                schema
-            ):
-                origins.extend(
-                    x for x in base.get_constraint_origins(schema)
-                    if x not in origins
-                )
 
-        return [self] if not origins else origins
+            # if this base applies
+            is_concrete = not base.is_non_concrete(schema)
+            if is_concrete and not base.get_delegated(schema):
+
+                # add all of its origins
+                for x in base.get_constraint_origins(schema):
+                    origins.add(x)
+
+        # if there no origin, return self
+        return [self] if not origins else list(origins)
```

Comment:
This is a refactor, me trying to understand if I can use this code for my case

Reply:
Probably not, right?

Reply:
No, it does something else (but that was not obvious to me before the refactor)